File: back/controller/domain.go
package controller

import (
	"chainqa_offchain_demo/models"
	"chainqa_offchain_demo/service"
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"
)

type DomainDTO struct {
	ApiUrl        ApiUrlDTO `json:"apiUrl"`        // API地址
	DomainName    string    `json:"domainName"`    // 数据域名称
	DomainMembers string    `json:"domainMembers"` // 数据域成员
	DomainPolicy  string    `json:"domainPolicy"`  // 数据域策略
	OrgId         string    `json:"orgId"`         // 组织ID
	Role          string    `json:"role"`          // 角色
}

func CreateDomainHandler(c *gin.Context) {
	fmt.Println("CreateDomainHandler")
	var req DomainDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "请求格式错误", err)
		return
	}
	createDomain(req, c)
}

func createDomain(req DomainDTO, c *gin.Context) {
	fmt.Println("createDomain")
	err := service.CreateDomain(req.ApiUrl.ContractName, req.ApiUrl.ChainServiceUrl, req.DomainName, req.DomainPolicy, req.OrgId)
	if err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "创建数据域失败", err)
		return
	}
	models.ResponseOK(c, "创建数据域成功", nil)
}

func UpdateDomainMetadataHandler(c *gin.Context) {
	fmt.Println("UpdateDomainMetadataHandler")
	var req DomainDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "请求格式错误", err)
		return
	}
	updateDomainMetadata(req, c)
}

func updateDomainMetadata(req DomainDTO, c *gin.Context) {
	fmt.Println("updateDomainMetadata")
	err := service.UpdateDomainMetadata(req.ApiUrl.ContractName, req.ApiUrl.ChainServiceUrl, req.DomainName, req.DomainMembers, req.DomainPolicy, req.OrgId)
	if err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "更新数据域元数据失败", err)
		return
	}
	models.ResponseOK(c, "更新数据域元数据成功", nil)
}

func QueryMyDomainsHandler(c *gin.Context) {
	fmt.Println("QueryMyDomainsHandler")
	var req DomainDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "请求格式错误", err)
		return
	}
	queryMyDomains(req, c)
}

func queryMyDomains(req DomainDTO, c *gin.Context) {
	fmt.Println("queryMyDomains")
	queryResult, err := service.QueryMyDomains(req.ApiUrl.ContractName, req.ApiUrl.ChainServiceUrl, req.OrgId)
	if err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "查询数据域失败", err)
		return
	}
	models.ResponseOK(c, "查询数据域成功", queryResult)
}

func QueryMyManagedDomainsHandler(c *gin.Context) {
	fmt.Println("QueryMyManagedDomainsHandler")
	var req DomainDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "请求格式错误", err)
		return
	}
	queryMyManagedDomains(req, c)
}

func queryMyManagedDomains(req DomainDTO, c *gin.Context) {
	fmt.Println("queryMyManagedDomains")
	queryResult, err := service.QueryMyManagedDomains(req.ApiUrl.ContractName, req.ApiUrl.ChainServiceUrl, req.OrgId)
	if err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "查询数据域失败", err)
		return
	}
	models.ResponseOK(c, "查询数据域成功", queryResult)
}

func QueryDomainInfoHandler(c *gin.Context) {
	fmt.Println("QueryDomainInfoHandler")
	var req DomainDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "请求格式错误", err)
		return
	}
	queryDomainInfo(req, c)
}

func queryDomainInfo(req DomainDTO, c *gin.Context) {
	fmt.Println("queryDomainInfo")
	queryResult, err := service.QueryDomainInfo(req.ApiUrl.ContractName, req.ApiUrl.ChainServiceUrl, req.DomainName, req.OrgId)
	if err != nil {
		models.ResponseError400(c, http.StatusBadRequest, "查询数据域失败", err)
		return
	}
	models.ResponseOK(c, "查询数据域成功", queryResult)
}

File: front/src/api/domain.js
import request from '@/api/request.js'
import { useApiStore } from '@/stores/api.js'
import { useUserStore } from '@/stores/user'

const baseUrl = '/domain'

/**
 * 创建数据域
 * @param {string} domainName - 数据域名称
 * @returns {Promise}
 */
export function createDomain(domainName) {
    const apiStore = useApiStore()
    const userStore = useUserStore()
    return request({
        url: baseUrl + '/createDomain',
        method: 'post',
        data: {
            apiUrl: {
                ipfsServiceUrl: apiStore.ipfsServiceUrl,
                contractName: apiStore.contractName,
                chainServiceUrl: apiStore.chainServiceUrl,
            },
            domainName: domainName,
            orgId: userStore.orgId,
            role: userStore.role,
        },
    })
}

/**
 * 更新数据域元数据
 * @param {string} domainName - 数据域名称
 * @param {string} domainMembers - 数据域成员
 * @param {string} domainPolicy - 数据域策略
 * @returns {Promise}
 */
export function updateDomainMetadata(domainName, domainMembers, domainPolicy) {
    const apiStore = useApiStore()
    const userStore = useUserStore()
    return request({
        url: baseUrl + '/updateDomainMetadata',
        method: 'post',
        data: {
            apiUrl: {
                ipfsServiceUrl: apiStore.ipfsServiceUrl,
                contractName: apiStore.contractName,
                chainServiceUrl: apiStore.chainServiceUrl,
            },
            domainName: domainName,
            domainMembers: domainMembers,
            domainPolicy: domainPolicy,
            orgId: userStore.orgId,
            role: userStore.role,
        },
    })
}

/**
 * 查询当前orgId下的数据域
 * @returns {Promise}
 */
export function queryMyDomains() {
    const apiStore = useApiStore()
    const userStore = useUserStore()
    return request({
        url: baseUrl + '/queryMyDomains',
        method: 'post',
        data: {
            apiUrl: {
                ipfsServiceUrl: apiStore.ipfsServiceUrl,
                contractName: apiStore.contractName,
                chainServiceUrl: apiStore.chainServiceUrl,
            },
            orgId: userStore.orgId,
            role: userStore.role,
        },
    })
}

/**
 * 查询当前orgId下管理的数据域
 * @returns {Promise}
 */
export function queryMyManagedDomains() {
    const apiStore = useApiStore()
    const userStore = useUserStore()
    return request({
        url: baseUrl + '/queryMyManagedDomains',
        method: 'post',
        data: {
            apiUrl: {
                ipfsServiceUrl: apiStore.ipfsServiceUrl,
                contractName: apiStore.contractName,
                chainServiceUrl: apiStore.chainServiceUrl,
            },
            orgId: userStore.orgId,
            role: userStore.role,
        },
    })
}

/**
 * 根据数据域名查询数据域详细信息
 * @param {string} domainName - 数据域名称
 * @returns {Promise}
 */
export function queryDomainInfo(domainName) {
    const apiStore = useApiStore()
    const userStore = useUserStore()
    return request({
        url: baseUrl + '/queryDomainInfo',
        method: 'post',
        data: {
            apiUrl: {
                ipfsServiceUrl: apiStore.ipfsServiceUrl,
                contractName: apiStore.contractName,
                chainServiceUrl: apiStore.chainServiceUrl,
            },
            domainName: domainName,
            orgId: userStore.orgId,
            role: userStore.role,
        },
    })
}

